### PR TITLE
Add uncompute functionality (WIP)

### PR DESCRIPTION
**Pull request by Manishearth**
As discussed in Taipei we plan to do animations in Stylo on the Rust side. For cascading properly, we need to "uncompute" these,
i.e. convert them into a cascadeable specified value which when computed gets us the same thing again.

This patch starts work on this. Before writing uncompute code for everything, I'd like to check that this is an acceptable amount of mako magic,
and that the general design is okay (to avoid having to rewrite everything once it's done).

Preliminary r? @SimonSapin @birtles

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13228)

### CODE REVIEW

**Comment by SimonSapin** [comment]
> For cascading properly, we need to "uncompute"

We talked about this last week, but I think I didn’t really understand it: why do we need to uncompute? It seems to me that that operation shouldn’t be necessary, but I haven’t thought it all through.

**Comment by Manishearth** [comment]
(discussed in IRC -- you need to recascade once you interpolate because of `!important`, so you need to go back to specified form)

**Comment by heycam** [comment]
Sorry for only getting a chance to look at this now.  I can understand wanting to eliminate the similar looking code we need to write to to_computed_value and uncompute, but looking at the couple of impl_to_computed_value uses you add in this patch, I find them quite hard to read.  To me it obfuscates the how these conversions are performed and how the specified and computed value types relate to each other.  I think I would prefer to see these written out by hand.

Another small comment: I never really liked the term "uncompute" much.  Can we call the function "to_specified_value" instead?  It feels a little odd for that function to be on the trait named ToComputedValue; maybe there's a better, more generic name for that trait we can use.

Why does ToComputedValue need to require Sized?

**Comment by Manishearth** [comment]
Finished all the remaining properties.

Debating on `to_specified_value` vs `from_computed_value` (since the function is implemented on the specified value).

Seems like it doesn't need `Sized`, not sure why I added it.

r? @heycam @birtles

File: components/style/values/computed/mod.rs
Comment (heycam) [inline components/style/values/computed/mod.rs]: Since you didn't need any context information for the `to_specified_value` implementations you've added, it's not clear whether we'll have any kind of useful context to pass in at the point we'll be calling `to_specified_value` from Gecko, and we no longer need it for the mako templates, let's remove it.

File: components/style/values/computed/mod.rs
Comment (heycam) [inline components/style/values/computed/mod.rs]: I like your suggestion of calling this `from_computed_value`.  Whether it makes sense to rename the trait to something like `ToFromComputedValue` or `ComputedValueConversion` I'm not sure.

File: components/style/values/computed/mod.rs
Comment (heycam) [inline components/style/values/computed/mod.rs]: recascading

File: components/style/properties/longhand/background.mako.rs
Comment (heycam) [inline components/style/properties/longhand/background.mako.rs]: (This renaming seems logically separate from the `to_specified_value` functionality so probably would have been better in a separate commit.)

File: components/style/values/computed/mod.rs
Comment (heycam) [inline components/style/values/computed/mod.rs]: A more general comment/question: is there an advantage to putting this function on `ToComputedValue` rather than on a new trait that the specified value types implement?  Is it just so that we can have two function definitions side by side, rather than potentially in separate modules/files?

File: components/style/properties/longhand/effects.mako.rs
Comment (heycam) [inline components/style/properties/longhand/effects.mako.rs]: Is `TranslateKind::Translate` the right thing to use here?  The specified value would round trip back to a computed value properly (since we always pass down the third component), but it wouldn't serialize properly.  I think this should either be `translate3d` always, or maybe based on whether `tz` is 0.

Incidentally, I don't think it's correct that the specified value loses whether it was a scale(), or scaleZ(), etc. when parsed.  That information should be preserved.  (It should also be preserved in the computed value, according to the spec.)

File: components/style/properties/longhand/inherited_text.mako.rs
Comment (heycam) [inline components/style/properties/longhand/inherited_text.mako.rs]: (Might be worth pulling out this type out to a LengthOrPercentageOrNormal type in values.rs later.)

File: components/style/values/specified/basic_shape.rs
Comment (heycam) [inline components/style/values/specified/basic_shape.rs]: Nit: I dunno if it's normal/preferred to line up the dots here like in `to_computed_value`...

**Review by heycam — CHANGES_REQUESTED** [review]

File: components/style/values/computed/mod.rs
Comment (Manishearth) [inline components/style/values/computed/mod.rs]: The primary use is still `ToComputedValue` so I think we can keep it that way.

Or `HasComputedValue`

Comment (Manishearth) [inline components/style/values/computed/mod.rs]: The specified->computed mapping is many-one; so the reverse trait can't exist. Well, it could exist, but with a trait type parameter, not an associated type.

**Comment by Manishearth** [comment]
Addressed. Split up the commits as well.

File: components/style/values/computed/mod.rs
Comment (heycam) [inline components/style/values/computed/mod.rs]: I see, that makes sense.

Comment (heycam) [inline components/style/values/computed/mod.rs]: OK, I leave it to you to decide.

**Comment by heycam** [comment]
@bors-servo delegate+

(In case you want to wait for review from @birtles.)

**Pull request merged**
